### PR TITLE
Added handling of unresolvable host names

### PR DESCRIPTION
**Pull request by vrozov**
@243826 Please review.

### CODE REVIEW

File: src/main/java/com/datatorrent/netlet/AbstractClient.java [hunk 1]
```diff
@@ -49,7 +49,7 @@
 
   public boolean isConnected()
   {
-    return key.isValid() && ((SocketChannel)key.channel()).isConnected();
+    return key != null && key.isValid() && ((SocketChannel)key.channel()).isConnected();
```
Comment (243826) [inline src/main/java/com/datatorrent/netlet/AbstractClient.java]: why would key be null?

Comment (vrozov) [inline src/main/java/com/datatorrent/netlet/AbstractClient.java]: In case `isConnected()` is called for the AbstractClient that did not connect successfully as in 'AbstractClientTest.testUnresolvedException()'.

File: src/main/java/com/datatorrent/netlet/DefaultEventLoop.java [hunk 2]
```diff
@@ -513,14 +512,18 @@ public String toString()
 
             });
           }
-        }
-        catch (IOException ie) {
-          l.handleException(ie, DefaultEventLoop.this);
+        } catch (UnresolvedAddressException e) {
+          logger.error("Inet Address {} is not resolvable.", address, e);
```
Comment (243826) [inline src/main/java/com/datatorrent/netlet/DefaultEventLoop.java]: error/warn/info - please do not do logging that that level. May be use trace or debug.

Comment (vrozov) [inline src/main/java/com/datatorrent/netlet/DefaultEventLoop.java]: Will it be OK to log this as the error as we need it for diagnostic in production deployment? We need the exception to be logged and if it is not logged by netlet, it will not be logged or handled anywhere outside of AbstractClient. Another approach will be to change `AbstractClient.handleException()` to log using error level. Netlet consumers may overwrite `handleException()` and log or handle exception the way they need.

Comment (243826) [inline src/main/java/com/datatorrent/netlet/DefaultEventLoop.java]: Will it be okay to log it as debug and turn on the debug logging for com.datatorrent.netlet in log4j.properties in the production application?

-
Chetan

Comment (PramodSSImmaneni) [inline src/main/java/com/datatorrent/netlet/DefaultEventLoop.java]: It may be difficult to do this as a default mechanism

Comment (243826) [inline src/main/java/com/datatorrent/netlet/DefaultEventLoop.java]: Yeah - probably one would need to muck with the Hadoop's log4j and it makes it less operable. How about using DTLoggerFactory?

Comment (tweise) [inline src/main/java/com/datatorrent/netlet/DefaultEventLoop.java]: +1 for logging this specific exception within netlet at error level

It is very hard to diagnose/support if we need to ask the user to change configuration to see the error. It is infrequent and important to see, using error level seems appropriate. Maybe add a comment to that effect?

File: src/main/java/com/datatorrent/netlet/AbstractClient.java (hunk 1 above)
Comment (tweise) [inline src/main/java/com/datatorrent/netlet/AbstractClient.java]: There is a test failure in jenkins:
{code}
java.lang.NullPointerException
	at com.datatorrent.netlet.AbstractClient.isConnected(AbstractClient.java:52)
	at com.datatorrent.netlet.DefaultEventLoopTest.testStop(DefaultEventLoopTest.java:249)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
{code}

Comment (243826) [inline src/main/java/com/datatorrent/netlet/AbstractClient.java]: It's known. The last open PR has a fix for it.

Comment (vrozov) [inline src/main/java/com/datatorrent/netlet/AbstractClient.java]: The last merged PR does not introduce the problem. The problem existed prior to the PR. The last PR relies on the fix in this PR.

Comment (243826) [inline src/main/java/com/datatorrent/netlet/AbstractClient.java]: Yes, you are right. I stand corrected.  :-)

File: src/main/java/com/datatorrent/netlet/DefaultEventLoop.java (hunk 2 above)
Comment (vrozov) [inline src/main/java/com/datatorrent/netlet/DefaultEventLoop.java]: My personal preference is to log all exception in `AbstractClient.handleException()` using error level. This frees library consumers from overwriting `handleException()` only to log the exception and in most cases it is the only action that the netlet consumer can take as `handleException()` is called on the `defaultEventLoop` thread. In cases where logging is not desired, it can be disabled by either overwriting "handleException' or by adjusting "com.datatorrent.netlet" logger level in log4j.

Many libraries, including Hadoop client library log errors and warnings, for example Hadoop will log a warning if it can't find or load native libraries.

If the concern is that the exception may be logged twice, it is a less severe issue than not logging or reporting error at all.

If everyone agrees on that approach, I will redo the commit.

Comment (PramodSSImmaneni) [inline src/main/java/com/datatorrent/netlet/DefaultEventLoop.java]: Can it be made a configuration setting of AbstractClient to log as error or debug, since it is anyway a specific implementation of handleException that AbstractClient is choosing to do.

Comment (tweise) [inline src/main/java/com/datatorrent/netlet/DefaultEventLoop.java]: +1 for logging in handleException at error level. It's one more option to suppress it for those that don't want to rely on log4j configuration.

Comment (243826) [inline src/main/java/com/datatorrent/netlet/DefaultEventLoop.java]: What about logging it as trace level and using DTLoggerFactory to enable it for com.datatorrent.netlet?

Also earlier PR did register the UncaughtExceptionHandler which logs it using error. So even if you do not log it here - you will see it in logs, right? I like that centralized logging as it open up an avenue to have the clients register their unhandled exception handler and have them do whatever they want in their.

-
Chetan

Comment (vrozov) [inline src/main/java/com/datatorrent/netlet/DefaultEventLoop.java]: No, the goal of logging UnresolvedAddressException is to log unresolvable InetAddress along with the exception, UncaughtExceptionHandler will not help as it is generic "catch all" case and it was introduced to log exceptions raised on the DefaultEventLoop to the log file instead of System.stderr where default UncaughtExceptionHandler outputs them. Additionally, we do want to route UnresolvedAddressException to AbstractClient.handleException, so it will not be propagated to UncaughtExceptionHandler.

I don't see how DTLoggerFactory will help, as we do want those exceptions to be logged by default and do not want to set "com.datatorrent.netlet" to trace level for all customer deployments.

**Comment by tweise** [comment]
@vrozov can you pull it in and proceed with the 1.2.1 release to unblock Apex 3.2.1?

**Comment by vrozov** [comment]
Working on the proposed changes. Will submit for review once they are ready.

**Comment by 243826** [comment]
If you guys are blocked... shade it in. Or log it at trace level and use dtlogger to enable the trace level for debugging your issues using netlet.

File: src/main/java/com/datatorrent/netlet/AbstractClient.java (hunk 1 above)
Comment (243826) [inline src/main/java/com/datatorrent/netlet/AbstractClient.java]: Vlad - do you want to send another PR for just this check so that the build passes consistently?

Comment (vrozov) [inline src/main/java/com/datatorrent/netlet/AbstractClient.java]: Yes, I am working on few test cases and will include the fix for the issue in the commit.

**Comment by vrozov** [comment]
Merged all changes to #49.
